common/nexus/payload_serializer.go

```diff
@@ -0,0 +1,221 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package nexus
+
+import (
+	"errors"
+	"fmt"
+	"mime"
+
+	"github.com/nexus-rpc/sdk-go/nexus"
+	commonpb "go.temporal.io/api/common/v1"
+)
+
+type payloadSerializer struct{}
+
+var errSerializer = errors.New("serializer error")
+
+// Deserialize implements nexus.Serializer.
+func (payloadSerializer) Deserialize(content *nexus.Content, v any) error {
+	payloadRef, ok := v.(**commonpb.Payload)
+	if !ok {
+		return fmt.Errorf("%w: cannot deserialize into %v", errSerializer, v)
+	}
+
+	payload := &commonpb.Payload{}
+	*payloadRef = payload
+	payload.Metadata = make(map[string][]byte)
+	payload.Data = content.Data
+
+	if !isStandardNexusContent(content) {
+		for k, v := range content.Header {
+			payload.Metadata[k] = []byte(v)
+		}
+		payload.Metadata["encoding"] = []byte("unknown/nexus-content")
+		return nil
+	}
+
+	contentType := content.Header.Get("type")
+	if contentType == "" {
+		payload.Metadata["encoding"] = []byte("binary/null")
+		return nil
+	}
+
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		return err
+	}
+
+	switch mediaType {
+	case "application/x-temporal-payload":
+		err := payload.Unmarshal(content.Data)
+		if err != nil {
+			return err
+		}
+		return nil
+	case "application/json":
+		if params["format"] == "protobuf" {
+			payload.Metadata["encoding"] = []byte("json/protobuf")
+			messageType := params["message-type"]
+			if messageType != "" {
+				payload.Metadata["messageType"] = []byte(messageType)
+			}
+		} else {
+			payload.Metadata["encoding"] = []byte("json/plain")
+		}
+	case "application/x-protobuf":
+		payload.Metadata["encoding"] = []byte("binary/protobuf")
+		messageType := params["message-type"]
+		if messageType != "" {
+			payload.Metadata["messageType"] = []byte(messageType)
+		}
+	case "application/octet-stream":
+		payload.Metadata["encoding"] = []byte("binary/plain")
+	default:
+		// Should be unreachable.
+		return fmt.Errorf("%w: standard content detection failed for %q", errSerializer, mediaType)
+	}
+	return nil
+}
+
+func isStandardNexusContent(content *nexus.Content) bool {
+	h := content.Header
+	// We assume that encoding is handled by the transport layer and the content is decoded.
+	delete(h, "encoding")
+	// Length can safely be ignored.
+	delete(h, "length")
+
+	if len(h) > 1 {
+		return false
+	}
+	contentType := h.Get("type")
+	if contentType == "" {
+		return len(h) == 0 && len(content.Data) == 0
```

Comment:
An absent content type + data is always asymmetric because it's more loosely defined inbound but we want to explicitly set content type outbound when data is present. So discussion of absent content-type + data only applies in one direction.